keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,623 @@
+---
+kep-number: 36
+title: Graduate Admission Webhooks to GA
+authors:
+  - "@mbohlool"
+owning-sig: sig-api-machinery
+reviewers:
+  - "@liggitt"
+approvers:
+  - "@liggitt"
+editor: TBD
+creation-date: 2019-01-27
+last-updated: 2019-01-29
+status: provisional
+see-also:
+  - [Admission Control Webhook Beta Design Doc](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/api-machinery/admission-control-webhooks.md)
+---
+
+# Admission Webhooks to GA
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Object selector](#object-selector)
+  * [Scope](#scope)
+  * [http for localhost](#http-for-localhost)
+  * [timeout configuration](#timeout-configuration)
+  * [Port configuration](#port-configuration)
+  * [Plumbing existing objects to delete admission](#plumbing-existing-objects-to-delete-admission)
+  * [Mutating Plugin ordering](#mutating-plugin-ordering)
+  * [Passing {Operation}Option to Webhook](#passing-operationoption-to-webhook)
+  * [AdmissionReview v1](#admissionreview-v1)
+* [V1 API](#v1-api)
+* [V1beta1 changes](#v1beta1-changes)
+* [Validations](#validations)
+* [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Post-GA tasks](#post-ga-tasks)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+Admission webhook is a way to extend kubernetes by putting hook on object creation/modification/deletion. Admission webhooks can mutate or validate the object. This feature has been Beta since Kubernetes 1.9. This document outline required steps to graduate it to GA.
+
+## Motivation
+
+Admission webhooks are currently widely used for extending kubernetes but has been in beta for 3 releases. Current set of feature requests and bug reports on the feature shows it is close to be stable and motivation of this KEP is to address those feature requests and bug reports to move this feature toward general availability (GA).
+
+### Goals
+
+Based on the user feedback, these are proposed changes to current feature to graduate it to GA:
+
+* Object Selector: Add a general selector for objects that a webhook can target.
+* Scope: Define the scope of the targeted objects, e.g. cluster vs. namespaced.
+* http for local host: relax https constrain for localhost.
+* timeout configuration: Add a configuration field to set the timeout of the webhook call.
+* port configuration: Add a configuration field to change the service port from the default 443.
```

Comment:
trying to remember why we couldn't just do this using URL.  Perhaps the detail explains

Reply:
some masters may not have kube-dns and you may not be able to point to services using urls reliably.

Reply:
What's the use case of supporting non-443 port in the configuration?

If the concern is that 443 is occupied on the pod, users can set the service's targetPort to a non-443 port.

Reply:
see below for types of services that can't make use of target port remapping